### PR TITLE
Jwineman/pi 955 http client

### PR DESCRIPTION
**Pull request by jwineman**
Ended up having to separate the http client entirely from the API client to satisfy the Magento2 code generation stuff.  Now we have:

`CF\API\HttpClientInterface`
`CF\API\DefaultHttpClient` (Guzzle 5 impl)

and `AbstractAPIClient` will implicitly depend on `DefaultHttpClient` in the interest of backwards compatibility but allow the injection of a new one through `setHttpClient()`.

### CODE REVIEW

File: src/API/DefaultHttpClient.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        return $response;
+    }
+
+    /**
+     * @param  Request $request
+     * @return GuzzleHttp\Message\RequestInterface $request
+     */
+    public function getGuzzleRequest(Request $request)
```
Comment (thellimist) [inline src/API/DefaultHttpClient.php]: instead of `getGuzzleRequest` let's change that to `createGuzzleRequest`

File: src/API/Plugin.php [hunk 2]
```diff
@@ -19,6 +21,14 @@ class Plugin extends Client
     const SETTING_EDITABLE_KEY = 'editable';
     const SETTING_MODIFIED_DATE_KEY = 'modified_on';
 
+    /**
+     * @param IntegrationInterface $integration
+     */
+    public function __construct(IntegrationInterface $integration)
+    {
+        parent::__construct($integration);
+    }
```
Comment (thellimist) [inline src/API/Plugin.php]: You don't need this function since it's already inherited

File: src/API/DefaultHttpClient.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    public function __construct($endpoint)
+    {
+        $this->client = new GuzzleHttp\Client(['base_url' => $endpoint]);
+    }
+
+    /**
+     * @param  Request $request
+     * @return Array $response
+     */
+    public function send(Request $request)
```
Comment (thellimist) [inline src/API/DefaultHttpClient.php]: add `throws RequestException` to php doc

File: src/API/AbstractAPIClient.php [hunk 4]
```diff
@@ -137,6 +112,21 @@ public function getPaginatedResults(Request $request, $response)
     }
 
     /**
+     * @param  Request $request
+     * @return Array $response
+     */
+    public function sendAndLog(Request $request)
```
Comment (thellimist) [inline src/API/AbstractAPIClient.php:118]: tests are missing for this function

File: src/API/Client.php [hunk 5]
```diff
@@ -11,6 +12,15 @@ class Client extends AbstractAPIClient
...
     const X_AUTH_EMAIL = 'X-Auth-Email';
 
+
+    /**
+     * @param IntegrationInterface $integration
+     */
+    public function __construct(IntegrationInterface $integration)
+    {
+        parent::__construct($integration);
+    }
```
Comment (thellimist) [inline src/API/Client.php]: don't need this function

File: src/API/Plugin.php (hunk 2 above)
Comment (jwineman) [reply src/API/Plugin.php]: Apparently its not.  Magento was giving me errors saying `__construct()` not found on `Client` and `Plugin`.

Comment (jwineman) [reply src/API/Plugin.php]: https://3v4l.org/oXBX0 never mind, must have been a bug in my impl.  Fat fingering the name of `__construct()` or something.

**Review by thellimist — APPROVED** [review]

**Pull request merged**